### PR TITLE
feat: replicas integration

### PR DESCRIPTION
**Pull request by nugaon**
Integration of #4453 

### Checklist

- [ ] I have read the [coding guide](https://github.com/ethersphere/bee/blob/master/CODING.md).
- [ ] My change requires a documentation update, and I have done it.
- [ ] I have added tests to cover my changes.
- [ ] I have filled out the description and linked the related issues.

### CODE REVIEW

File: pkg/replicas/replicas.go [hunk 1]
```diff
@@ -125,21 +112,30 @@ func (rr *replicator) replicas() {
 }
 
 // add inserts the soc replica into a replicator so that addresses are balanced
-func (rr *replicator) add(r *replica, d uint8) (depth int, rank int) {
-	if d == 0 {
+func (rr *replicator) add(r *replica, rLevel redundancy.Level) (depth int, rank int) {
```
Comment (zelig) [inline pkg/replicas/replicas.go:115]: these changes do not improve the code,

File: pkg/file/redundancy/level.go [hunk 2]
```diff
@@ -135,3 +150,19 @@ var encParanoidEt = newErasureTable(
 		55, 51, 48, 44, 39, 35, 30, 24,
 	},
 )
+
```
Comment (zelig) [inline pkg/file/redundancy/level.go:153]: These changes expose a lot of convoluted internals to public which is a sign that they are not at the right place. 
I think this should remain as per the original, honestly.

File: pkg/file/redundancy/level.go [hunk 3]
```diff
@@ -82,6 +82,21 @@ func (l Level) GetMaxEncShards() int {
 	return (swarm.Branches - p) / 2
 }
 
+// GetReplicaCount returns back the dispersed replica number
```
Comment (zelig) [inline pkg/file/redundancy/level.go:85]: These changes expose a lot of convoluted internals to public which is a sign that they are not at the right place.
I think this should remain as per the original, honestly.

File: pkg/file/pipeline/hashtrie/hashtrie.go [hunk 4]
```diff
@@ -21,20 +25,29 @@ var (
 const maxLevel = 8
 
 type hashTrieWriter struct {
+	ctx                    context.Context // context for put function of dispersed replica chunks
```
Comment (zelig) [inline pkg/file/pipeline/hashtrie/hashtrie.go:28]: context should not be a field if possible

File: pkg/file/pipeline/hashtrie/hashtrie.go [hunk 5]
```diff
@@ -21,20 +25,29 @@ var (
...
 	pipelineFn             pipeline.PipelineFunc
 	rParams                redundancy.IParams
 	parityChunkFn          redundancy.ParityChunkCallback
-	chunkCounters          []uint8 // counts the chunk references in intermediate chunks. key is the chunk level.
-	effectiveChunkCounters []uint8 // counts the effective  chunk references in intermediate chunks. key is the chunk level.
-	maxChildrenChunks      uint8   // maximum number of chunk references in intermediate chunks.
+	chunkCounters          []uint8        // counts the chunk references in intermediate chunks. key is the chunk level.
+	effectiveChunkCounters []uint8        // counts the effective  chunk references in intermediate chunks. key is the chunk level.
+	maxChildrenChunks      uint8          // maximum number of chunk references in intermediate chunks.
+	replicaPutter          storage.Putter // putter to save dispersed replicas of the root chunk
```
Comment (zelig) [inline pkg/file/pipeline/hashtrie/hashtrie.go:39]: if only used for replicas then this should be right away initialised  with `replicas.NewPutter`

File: pkg/file/pipeline/hashtrie/hashtrie.go [hunk 6]
```diff
@@ -245,5 +259,19 @@ func (h *hashTrieWriter) Sum() ([]byte, error) {
...
-	return data[swarm.SpanSize:], nil
+	rootHash := data[swarm.SpanSize:]
+
+	// save disperse replicas of the root chunk
+	if h.rParams.Level() != redundancy.NONE {
+		rootData, err := h.rParams.GetRootData()
+		if err != nil {
+			return nil, err
+		}
+		putter := replicas.NewPutter(h.replicaPutter)
```
Comment (zelig) [inline pkg/file/pipeline/hashtrie/hashtrie.go:270]: this should already be part of the `hashTrieWriter` inititialisation

File: pkg/file/pipeline/hashtrie/hashtrie.go [hunk 7]
```diff
@@ -245,5 +259,19 @@ func (h *hashTrieWriter) Sum() ([]byte, error) {
...
+		err = putter.Put(h.ctx, swarm.NewChunk(swarm.NewAddress(rootHash), rootData))
```
Comment (zelig) [inline pkg/file/pipeline/hashtrie/hashtrie.go:271]: here we should not use context from the `hashTrieWriter`. Either `Sum` should be extended with a context or in fact the whole integration is possibly at the wrong place. I think it should happen after and where Sum is called. There the putter/getter used underlyingly is known so its easy to then just wrap the putter in the replicas.Putter and use that to Put the chunk gotten from the getter.

**Review by zelig — CHANGES_REQUESTED** [review]
I have reservations as to the content of this PR. 
It should really be super simple, the fact that it is not indicates design problem.

Also I am missing the replicas integration into the joiner. which should be trivial.

File: pkg/file/pipeline/hashtrie/hashtrie.go (hunk 4 above)
Comment (nugaon) [reply pkg/file/pipeline/hashtrie/hashtrie.go:28]: hashtrie's Sum() method should satisfy pipeline interface that does not define context.
context is needed to put SOC replica.

File: pkg/file/pipeline/hashtrie/hashtrie.go (hunk 7 above)
Comment (nugaon) [reply pkg/file/pipeline/hashtrie/hashtrie.go:272]: the redundancy cache is available in the hashtrie package from which the root chunk data is retrieved, so I would argue whether the integration is a great place.

**Comment by nugaon** [comment]
[data race error](https://github.com/ethersphere/bee/actions/runs/7214662399/job/19657491688?pr=4492) will be fixed by the new refactored test cases in #4507

File: pkg/file/redundancy/level.go (hunk 3 above)
Comment (nugaon) [reply pkg/file/redundancy/level.go:101]: `GetReplicaCount` is called in hashtrie_test.go and IMO it is like erasure tables.
I moved indexBases back to replicas packages as we discussed offline.

**Pull request merged**